File: examples/cpp/Beta/mlsd_parser.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>
#include <string>

#include "depthai/beta/datatypes.hpp"
#include "depthai/beta/nodes.hpp"
#include "depthai/depthai.hpp"

int main() {
    dai::Pipeline pipeline;

    const std::string modelSlug = "luxonis/m-lsd:512x512";
    dai::NNModelDescription modelDescription{modelSlug};
    modelDescription.platform = pipeline.getDefaultDevice()->getPlatformAsString();
    dai::NNArchive modelArchive(dai::getModelFromZoo(modelDescription));

    auto cameraNode = pipeline.create<dai::node::Camera>()->build();
    auto neuralNetwork = pipeline.create<dai::node::NeuralNetwork>()->build(cameraNode, modelArchive);
    auto parserNode = pipeline.create<dai::beta::node::MLSDParser>()->build(neuralNetwork->out, modelArchive);

    auto frameQueue = neuralNetwork->passthrough.createOutputQueue();
    auto outputQueue = parserNode->out.createOutputQueue();
    auto configQueue = parserNode->inputConfig.createInputQueue();
    auto config = parserNode->initialConfig;

    pipeline.start();
    std::cout << "Controls: '+' increase score threshold, '-' decrease it, 'q' quit." << std::endl;

    while(pipeline.isRunning()) {
        auto frameMessage = frameQueue->get<dai::ImgFrame>();
        auto parserOutput = outputQueue->get<dai::beta::Lines>();
        cv::Mat frame = frameMessage->getCvFrame();
        for(const auto& line : parserOutput->lines) {
            const cv::Point startPoint(static_cast<int>(line.startPoint.x * frame.cols), static_cast<int>(line.startPoint.y * frame.rows));
            const cv::Point endPoint(static_cast<int>(line.endPoint.x * frame.cols), static_cast<int>(line.endPoint.y * frame.rows));
            cv::line(frame, startPoint, endPoint, cv::Scalar(0, 255, 0), 2);
        }

        cv::imshow("MLSDParser", frame);
        const int key = cv::waitKey(1);
        if(key == 'q') break;
        if(key == '+') {
            config->scoreThreshold = std::min(1.0f, config->scoreThreshold + 0.1f);
            configQueue->send(config);
            std::cout << "Score threshold: " << config->scoreThreshold << std::endl;
        } else if(key == '-') {
            config->scoreThreshold = std::max(0.0f, config->scoreThreshold - 0.1f);
            configQueue->send(config);
            std::cout << "Score threshold: " << config->scoreThreshold << std::endl;
        }
    }

    return 0;
}

File: examples/python/Beta/mlsd_parser.py
#!/usr/bin/env python3

import cv2
import depthai as dai


def main() -> None:
    with dai.Pipeline() as pipeline:
        modelSlug = "luxonis/m-lsd:512x512"
        modelDescription = dai.NNModelDescription(
            modelSlug,
            platform=pipeline.getDefaultDevice().getPlatformAsString(),
        )
        modelArchive = dai.NNArchive(dai.getModelFromZoo(modelDescription))

        cameraNode = pipeline.create(dai.node.Camera).build()
        neuralNetwork = pipeline.create(dai.node.NeuralNetwork).build(cameraNode, modelArchive)
        parserNode = pipeline.create(dai.beta.node.MLSDParser).build(
            neuralNetwork.out,
            modelArchive,
        )

        frameQueue = neuralNetwork.passthrough.createOutputQueue()
        outputQueue = parserNode.out.createOutputQueue()
        configQueue = parserNode.inputConfig.createInputQueue()
        config = parserNode.initialConfig

        pipeline.start()
        print("Controls: '+' increase score threshold, '-' decrease it, 'q' quit.")

        while pipeline.isRunning():
            frameMessage = frameQueue.get()
            parserOutput = outputQueue.get()
            frame = frameMessage.getCvFrame()
            for line in parserOutput.lines:
                startPoint = (
                    int(line.startPoint.x * frame.shape[1]),
                    int(line.startPoint.y * frame.shape[0]),
                )
                endPoint = (
                    int(line.endPoint.x * frame.shape[1]),
                    int(line.endPoint.y * frame.shape[0]),
                )
                cv2.line(frame, startPoint, endPoint, (0, 255, 0), 2)


            cv2.imshow("MLSDParser", frame)
            key = cv2.waitKey(1)
            if key == ord("q"):
                break
            if key == ord("+"):
                config.scoreThreshold = min(1.0, config.scoreThreshold + 0.1)
                configQueue.send(config)
                print(f"Score threshold: {config.scoreThreshold:.1f}")
            elif key == ord("-"):
                config.scoreThreshold = max(0.0, config.scoreThreshold - 0.1)
                configQueue.send(config)
                print(f"Score threshold: {config.scoreThreshold:.1f}")


if __name__ == "__main__":
    main()
